Sync MK3<->MK4 MMU2

DRracer opened the pull request:
This is a collection of commits towards unifying the code base on MK3 and MK4.

Best reviewed per commit as there are many small changes scattered around the MMU-related code base.

Commit:
Sync MK3<->MK4 MMU2 protocol logic

Commit:
Sync MK3<->MK4 MMU2 crc

Commit:
Clean-up ProgressCodes and ErrorCodes parameters throughout the code

Commit:
Sync MK3<->MK4 MMU2 progress converter

Commit:
Sync MK3<->MK4 MMU2 log

Commit:
Sync MK3<->MK4 MMU2 marlin interface

Commit:
Sync MK3<->MK4 MMU2 marlin macros

Commit:
Sync MK3<->MK4 MMU2 reporting

Commit:
Sync MK3<->MK4 MMU2 mmu2_mk4.h

Commit:
Sync MK3<->MK4 MMU2 mmu2_mk4.cpp formatting

Commit:
Sync MK3<->MK4 MMU2 TryLoadUnloadReporter

Commit:
Sync MK3<->MK4 MMU2 Tune, Buttons, PowerPanic hooks refactoring

gudnimg commented on Firmware/mmu2.cpp:
```diff
@@ -295,42 +285,23 @@ bool MMU2::VerifyFilamentEnteredPTFE() {
-
-    Disable_E0();
-    TryLoadUnloadProgressbarEcho();
-    TryLoadUnloadProgressbarDeinit();
```
I think these two lines are removed by mistake. 

Echo will log the result onto serial. 

Deinit will signal to the LCD code that the next status message should be rendered from the beginning.

DRracer replied:
Good point, you are right, I missed the serial dump... I'll probably push it into the destructor of the `TryLoadUnloadReporter`

DRracer commented:
rebased onto MK3

Commit:
Return try-load reporting to serial

Commit:
Unify MMU operations' workflow

gudnimg commented on Firmware/mmu2_protocol_logic.cpp:
```diff
@@ -9,36 +9,43 @@
 ///
 /// Currently, this is not possible to perform automatically at compile time with the existing languages/translations infrastructure.
 /// To save space a "dumb" solution was chosen + a few static_assert checks in errors_list.h preventing the code from compiling when the string doesn't match.
+/// -----
+/// On Buddy FW we should improve the error screen to be able to print formatted strings
```
Does this TODO need to be in the code? I feel it would be better to create a JIRA ticket for this as it only relates to the 32-bit.

gudnimg commented on Firmware/mmu2_reporting.h:
```diff
@@ -2,12 +2,19 @@
 enum CommandInProgress : uint8_t {
     NoCommand = 0,
-    CutFilament = 'C',
+    CutFilament = 'K',
```
Nice 👍

gudnimg commented on Firmware/mmu2.cpp:
```diff
@@ -295,42 +284,23 @@ bool MMU2::VerifyFilamentEnteredPTFE() {
         }
     }
-
-    Disable_E0();
```
It would be nice to keep this `Disable_E0();` It is relevant for single color prints. Tx/Tc will execute this code before heating is complete. So disabling the E-motor sooner will be a better user experience. The printer is more silent while heating.

gudnimg commented:
Code changes look great, just minor comments above.
Will test this PR on hardware tomorrow before approving👍 


**EDIT:**

I didn't notice before but it seems this PR is reverting the improvements in this PR: https://github.com/prusa3d/Prusa-Firmware/pull/4080

Note the changes I made to `TryLoadUnloadProgressbarDeinit`, `TryLoadUnloadProgressbarInit` and `TryLoadUnloadProgressbar` in 3.13.2 to keep the status screen rendering without problems. This solved a few problems with the LCD rendering.

It looks like this PR code is based on the code before PR 4080

gudnimg commented on Firmware/mmu2_reporting.cpp:
```diff
@@ -295,30 +295,32 @@ void ReportProgressHook(CommandInProgress cip, ProgressCode ec) {
+    // Set the cursor position each time in case some other
+    // part of the firmware changes the cursor position
+    lcd_putc_at(col, 3, sensorState ? '-' : LCD_STR_SOLID_BLOCK[0]);
     lcd_reset_status_message_timeout();
```
Note this line is still missing `lcd_reset_status_message_timeout()`

I'm not sure if this should be used in `TryLoadUnloadReporter::Render()`

gudnimg replied:
Here we should use `if (!lcd_update_enabled) lcdui_print_status_line();` this is hack to allow showing the progress bar in fullscreens like "Loading Filament 1".

gudnimg commented:
Compiler warnings
![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/d62a3f17-2bd2-4dc9-8419-a57b122c396d)

This below fixes it:

![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/03104bbc-01d9-46d6-b387-989f51fa8502)

gudnimg commented on Firmware/mmu2.cpp:
```diff
@@ -295,42 +284,23 @@ bool MMU2::VerifyFilamentEnteredPTFE() {
-                TryLoadUnloadProgressbar(lcd_cursor_col++, fsensorStateLCD);
-                fsensorStateLCD = 0;      // Clear temporary bit
-            }
+            filament_inserted = filament_inserted && (WhereIsFilament() == FilamentState::AT_FSENSOR);
```
The logic is flipped compared to previous solution in 3.13.2.

This may be a good change in scenario when the filament sensor is reporting a false positive 0 value. 3.13.2 would in that case always fail the test (from one sample). But with this change, all the samples for a given LCD pixel must now see 0 values from the filament sensor. Could be useful for users who have the filament sensor flickering occasionally if sampled very fast.

gudnimg committed:
8-bit proposed changes

The status line code is not nice, but we need to work around it
so the status line rendering works correctly

This commit mostly reapplies the code
from 3.13.2

Also fixes compiler warnings

gudnimg commented:
Just for reference: a PR with my proposed changes https://github.com/DRracer/Prusa-Firmware/pull/80

Commit:
Fixup from review
